File: backend/app/routes/candidates.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List
from app.database import get_db
from app.models.candidate import Candidate
from app.models.user import User
from app.schemas import CandidateCreate, CandidateUpdate, CandidateResponse, ApiResponse
from app.auth import get_current_active_user, require_role

router = APIRouter()


@router.get("/", response_model=ApiResponse)
async def get_all_candidates(
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db)
):
    candidates = db.query(Candidate).order_by(Candidate.created_at.desc()).all()
    
    return ApiResponse(
        success=True,
        data={
            "candidates": [{
                "id": str(c.id),
                "name": c.name,
                "email": c.email,
                "phone": c.phone,
                "location": c.location,
                "detected_role": c.detected_role,
                "target_role": c.target_role,
                "summary": c.summary,
                "education_degree": c.education_degree,
                "education_institution": c.education_institution,
                "education_graduation_year": c.education_graduation_year,
                "github_url": c.github_url,
                "linkedin_url": c.linkedin_url,
                "portfolio_url": c.portfolio_url,
                "is_development_data": c.is_development_data,
                "created_at": c.created_at.isoformat(),
                "updated_at": c.updated_at.isoformat()
            } for c in candidates]
        }
    )


@router.get("/{candidate_id}", response_model=ApiResponse)
async def get_candidate(
    candidate_id: str,
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db)
):
    candidate = db.query(Candidate).filter(Candidate.id == candidate_id).first()
    if not candidate:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Candidate not found"
        )
    
    return ApiResponse(
        success=True,
        data={
            "candidate": {
                "id": str(candidate.id),
                "name": candidate.name,
                "email": candidate.email,
                "phone": candidate.phone,
                "location": candidate.location,
                "detected_role": candidate.detected_role,
                "target_role": candidate.target_role,
                "summary": candidate.summary,
                "education_degree": candidate.education_degree,
                "education_institution": candidate.education_institution,
                "education_graduation_year": candidate.education_graduation_year,
                "github_url": candidate.github_url,
                "linkedin_url": candidate.linkedin_url,
                "portfolio_url": candidate.portfolio_url,
                "is_development_data": candidate.is_development_data,
                "created_at": candidate.created_at.isoformat(),
                "updated_at": candidate.updated_at.isoformat()
            }
        }
    )


@router.post("/", response_model=ApiResponse)
async def create_candidate(
    candidate_data: CandidateCreate,
    current_user: User = Depends(require_role("RECRUITER", "ADMIN")),
    db: Session = Depends(get_db)
):
    if not candidate_data.name:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Name is required"
        )
    
    new_candidate = Candidate(**candidate_data.model_dump())
    db.add(new_candidate)
    db.commit()
    db.refresh(new_candidate)
    
    return ApiResponse(
        success=True,
        data={
            "candidate": {
                "id": str(new_candidate.id),
                "name": new_candidate.name,
                "email": new_candidate.email,
                "phone": new_candidate.phone,
                "location": new_candidate.location,
                "detected_role": new_candidate.detected_role,
                "target_role": new_candidate.target_role,
                "summary": new_candidate.summary,
                "education_degree": new_candidate.education_degree,
                "education_institution": new_candidate.education_institution,
                "education_graduation_year": new_candidate.education_graduation_year,
                "github_url": new_candidate.github_url,
                "linkedin_url": new_candidate.linkedin_url,
                "portfolio_url": new_candidate.portfolio_url,
                "is_development_data": new_candidate.is_development_data,
                "created_at": new_candidate.created_at.isoformat(),
                "updated_at": new_candidate.updated_at.isoformat()
            }
        }
    )


@router.put("/{candidate_id}", response_model=ApiResponse)
async def update_candidate(
    candidate_id: str,
    candidate_data: CandidateUpdate,
    current_user: User = Depends(require_role("RECRUITER", "ADMIN")),
    db: Session = Depends(get_db)
):
    candidate = db.query(Candidate).filter(Candidate.id == candidate_id).first()
    if not candidate:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Candidate not found"
        )
    
    # Update fields
    update_data = candidate_data.model_dump(exclude_unset=True)
    for field, value in update_data.items():
        setattr(candidate, field, value)
    
    db.commit()
    db.refresh(candidate)
    
    return ApiResponse(
        success=True,
        data={
            "candidate": {
                "id": str(candidate.id),
                "name": candidate.name,
                "email": candidate.email,
                "phone": candidate.phone,
                "location": candidate.location,
                "detected_role": candidate.detected_role,
                "target_role": candidate.target_role,
                "summary": candidate.summary,
                "education_degree": candidate.education_degree,
                "education_institution": candidate.education_institution,
                "education_graduation_year": candidate.education_graduation_year,
                "github_url": candidate.github_url,
                "linkedin_url": candidate.linkedin_url,
                "portfolio_url": candidate.portfolio_url,
                "is_development_data": candidate.is_development_data,
                "created_at": candidate.created_at.isoformat(),
                "updated_at": candidate.updated_at.isoformat()
            }
        }
    )


@router.delete("/{candidate_id}", response_model=ApiResponse)
async def delete_candidate(
    candidate_id: str,
    current_user: User = Depends(require_role("RECRUITER", "ADMIN")),
    db: Session = Depends(get_db)
):
    candidate = db.query(Candidate).filter(Candidate.id == candidate_id).first()
    if not candidate:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Candidate not found"
        )
    
    db.delete(candidate)
    db.commit()
    
    return ApiResponse(
        success=True,
        message="Candidate deleted successfully"
    )

File: server-legacy/src/routes/candidates.ts
import { Router, Response, NextFunction } from 'express';
import { query } from '../database/index.js';
import { authenticate, AuthRequest, requireRole } from '../middleware/auth.js';
import { ValidationError } from '../middleware/errorHandler.js';

const router = Router();

// Get all candidates
router.get('/', authenticate, async (req: AuthRequest, res: Response, next: NextFunction) => {
  try {
    const result = await query(
      'SELECT * FROM candidates ORDER BY created_at DESC'
    );

    res.json({
      success: true,
      data: result.rows
    });
  } catch (error) {
    next(error);
  }
});

// Get single candidate
router.get('/:id', authenticate, async (req: AuthRequest, res: Response, next: NextFunction) => {
  try {
    const { id } = req.params;
    const result = await query('SELECT * FROM candidates WHERE id = $1', [id]);

    if (result.rows.length === 0) {
      throw new ValidationError('Candidate not found');
    }

    res.json({
      success: true,
      data: result.rows[0]
    });
  } catch (error) {
    next(error);
  }
});

// Create candidate
router.post('/', authenticate, requireRole(['RECRUITER', 'ADMIN']), async (req: AuthRequest, res: Response, next: NextFunction) => {
  try {
    const {
      name,
      email,
      phone,
      location,
      detected_role,
      target_role,
      summary,
      education_degree,
      education_institution,
      education_graduation_year,
      github_url,
      linkedin_url,
      portfolio_url
    } = req.body;

    if (!name) {
      throw new ValidationError('Name is required');
    }

    const result = await query(
      `INSERT INTO candidates (
        name, email, phone, location, detected_role, target_role, summary,
        education_degree, education_institution, education_graduation_year,
        github_url, linkedin_url, portfolio_url
      ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13)
      RETURNING *`,
      [
        name, email, phone, location, detected_role, target_role, summary,
        education_degree, education_institution, education_graduation_year,
        github_url, linkedin_url, portfolio_url
      ]
    );

    res.status(201).json({
      success: true,
      data: result.rows[0]
    });
  } catch (error) {
    next(error);
  }
});

// Update candidate
router.put('/:id', authenticate, requireRole(['RECRUITER', 'ADMIN']), async (req: AuthRequest, res: Response, next: NextFunction) => {
  try {
    const { id } = req.params;
    const {
      name,
      email,
      phone,
      location,
      detected_role,
      target_role,
      summary,
      education_degree,
      education_institution,
      education_graduation_year,
      github_url,
      linkedin_url,
      portfolio_url
    } = req.body;

    const result = await query(
      `UPDATE candidates SET
        name = COALESCE($1, name),
        email = COALESCE($2, email),
        phone = COALESCE($3, phone),
        location = COALESCE($4, location),
        detected_role = COALESCE($5, detected_role),
        target_role = COALESCE($6, target_role),
        summary = COALESCE($7, summary),
        education_degree = COALESCE($8, education_degree),
        education_institution = COALESCE($9, education_institution),
        education_graduation_year = COALESCE($10, education_graduation_year),
        github_url = COALESCE($11, github_url),
        linkedin_url = COALESCE($12, linkedin_url),
        portfolio_url = COALESCE($13, portfolio_url),
        updated_at = CURRENT_TIMESTAMP
      WHERE id = $14
      RETURNING *`,
      [
        name, email, phone, location, detected_role, target_role, summary,
        education_degree, education_institution, education_graduation_year,
        github_url, linkedin_url, portfolio_url, id
      ]
    );

    if (result.rows.length === 0) {
      throw new ValidationError('Candidate not found');
    }

    res.json({
      success: true,
      data: result.rows[0]
    });
  } catch (error) {
    next(error);
  }
});

// Delete candidate
router.delete('/:id', authenticate, requireRole(['RECRUITER', 'ADMIN']), async (req: AuthRequest, res: Response, next: NextFunction) => {
  try {
    const { id } = req.params;
    const result = await query('DELETE FROM candidates WHERE id = $1 RETURNING id', [id]);

    if (result.rows.length === 0) {
      throw new ValidationError('Candidate not found');
    }

    res.json({
      success: true,
      message: 'Candidate deleted successfully'
    });
  } catch (error) {
    next(error);
  }
});

export { router as candidatesRouter };
